app/src/main/java/com/duckduckgo/app/browser/pageloadpixel/firstpaint/PagePaintedHandler.kt

```diff
@@ -0,0 +1,96 @@
+/*
+ * Copyright (c) 2024 DuckDuckGo
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.duckduckgo.app.browser.pageloadpixel.firstpaint
+
+import android.webkit.WebView
+import com.duckduckgo.app.browser.pageloadpixel.PageLoadedSites.Companion.sites
+import com.duckduckgo.app.di.AppCoroutineScope
+import com.duckduckgo.browser.api.WebViewVersionProvider
+import com.duckduckgo.common.utils.ConflatedJob
+import com.duckduckgo.common.utils.DispatcherProvider
+import com.duckduckgo.common.utils.UriString
+import com.duckduckgo.common.utils.device.DeviceInfo
+import com.duckduckgo.di.scopes.AppScope
+import com.squareup.anvil.annotations.ContributesBinding
+import javax.inject.Inject
+import kotlin.coroutines.resume
+import kotlin.coroutines.suspendCoroutine
+import kotlin.math.roundToLong
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.launch
+import kotlinx.coroutines.withContext
+import timber.log.Timber
+
+interface PagePaintedHandler {
+    operator fun invoke(
+        webView: WebView,
+        url: String,
+    )
+}
+
+@ContributesBinding(AppScope::class)
+class RealPagePaintedHandler @Inject constructor(
+    private val deviceInfo: DeviceInfo,
+    private val webViewVersionProvider: WebViewVersionProvider,
+    private val dao: PagePaintedPixelDao,
+    @AppCoroutineScope private val appCoroutineScope: CoroutineScope,
+    private val dispatcherProvider: DispatcherProvider,
+) : PagePaintedHandler {
+
+    private val job = ConflatedJob()
+
+    override operator fun invoke(
+        webView: WebView,
+        url: String,
+    ) {
+        job += appCoroutineScope.launch(dispatcherProvider.io()) {
```

Comment:
Why do we want to cancel the previous job?

Reply:
If we hit a scenario where `onPageFinished()` is called multiple times, or there are multiple quick redirects for example, i'm happy to prioritise performance and drop a pixel here or there. Probably fine to not cancel, but seems slightly safer to.